packages/sanity/src/core/config/components/useMiddlewareComponents.ts

```diff
@@ -1,31 +1,39 @@
-import {ComponentType, createElement, useMemo} from 'react'
+/* eslint-disable @typescript-eslint/ban-types */
+
+import {ComponentType, createElement, Fragment, ReactElement, useMemo} from 'react'
 import {useSource} from '../../studio'
 import {PluginOptions} from '../types'
 
-interface MiddlewareProps<T> {
-  renderDefault: (props: T) => React.ReactElement
-}
+const emptyRender = () => createElement(Fragment)
 
-function _createMiddlewareComponent<T extends {renderDefault: (props: T) => React.ReactElement}>(
+function _createMiddlewareComponent<T extends {}>(
+  defaultComponent: ComponentType<T>,
   middlewareComponents: ComponentType<T>[]
 ): ComponentType<T> {
-  return (_props: T) => {
-    const _defaultItem = (props: T) => createElement(middlewareComponents[0], props)
-
-    let next = _defaultItem
-
-    for (const middleware of middlewareComponents.slice(1)) {
-      const defaultRender = next
-
-      next = (props) => {
-        return createElement(middleware, {
-          ...props,
-          renderDefault: defaultRender,
-        })
-      }
+  return (outerProps: T) => {
+    // This is the inner "layer" of the middleware chain
+    // Here we render the _default_ component (typically Sanity's component)
+    let next: (p: T) => ReactElement = (props: T) => createElement(defaultComponent, props)
+
+    // Since the middleware array is actually a middleware _stack_ data structure, we are
+    // essentially looping backwards here. This makes it possible to define the _next_ layer for
+    // each layer.
+    for (const middleware of middlewareComponents) {
+      // As we progress through the chain, the meaning of "renderDefault" changes.
+      // At a given layer in the chain, the _next_ layer is the "default".
+      const renderDefault = next
+
+      // Here we replace `next` so that the _previous_ layer may use this as its _next_.
+      next = (props) => createElement(middleware, {...props, renderDefault})
     }
 
-    return next(_props)
+    return next({
+      ...outerProps,
+      // NOTE: it's safe to pass the empty render function, since it'll be overwritten in the next step (above).
+      // NOTE: it's important that the default component does not use `renderDefault`, since it will
+      // get the `emptyRender` callback will be passed when the middleware stack is empty.
+      renderDefault: emptyRender,
+    })
   }
```

Comment:
I think it was good to get these comments added, to document what's going on here.